nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -793,6 +796,61 @@ impl DataStore {
         Ok(updated)
     }
 
+    pub async fn instance_resize(
+        &self,
+        opctx: &OpContext,
+        instance_id: InstanceUuid,
+        cpus: InstanceCpuCount,
+        memory: ByteCount,
+    ) -> UpdateResult<()> {
+        use db::model::InstanceState as DbInstanceState;
+        use db::schema::instance::dsl;
+
+        const ALLOWED_RESIZE_STATES: &[DbInstanceState] =
+            &[DbInstanceState::NoVmm, DbInstanceState::Failed];
+
+        // Use check_if_exists to distinguish cases where the instance doesn't
+        // exist at all from cases where it exists but is in the wrong state.
+        let instance_id = instance_id.into_untyped_uuid();
+        let updated = diesel::update(dsl::instance)
+            .filter(dsl::time_deleted.is_null())
+            .filter(dsl::id.eq(instance_id))
+            .filter(dsl::state.eq_any(ALLOWED_RESIZE_STATES))
```

Comment:
I was going to ask if we should also check for the `propolis_id` being null here, but I suppose we can rely on the check constraint that the instance can only be in `NoVmm` if it's null. Maybe worth commenting here?